### ISSUE TITLE
Intermittent non-repeatable HGVSDataNotAvailableError: No sequence available for ...

### ISSUE THREAD

**Issue by jfreidin** [biocommons/hgvs#519]
Hi, I'm running a local docker-based uta and am seeing the following with a frequency of ~1/1000 fetches.
I'm typically running multiple hgvs processes at once, not sure if that's relevant.
I have no idea what the root cause might be or how to debug it since I can't reproduce it at will.
For a workaround I've implemented retry logic, which has always succeeded.
Any help identifying the root cause would be appreciated.
I've seen this fail in multiple code paths that rely on fetch_seq and for different transcripts.
An example:

Traceback (most recent call last):
  File "pipecycle/cohort.py", line 79, in count_reads_bam_file
    mon_cohort.count_reads_bam_file(bam_fofn_row, alteration, is_control, mon_output)
  File "pipecycle/cohort.py", line 61, in count_reads_bam_file
    context_bases=self.context_bases)
  File "pipecycle/readclassifier.py", line 9, in __init__
    super(ReadClassifier, self).__init__(gene, ref_seq_transcript, fmi_cds_effect, context_bases, is_print_alignment)
  File "pipecycle/shortvariant.py", line 22, in __init__
    self.variant_p = assembly_mapper.c_to_p(self.variant_c)
  File "/Users/jfreidin/.pyenv/versions/pipecycle/lib/python2.7/site-packages/hgvs/assemblymapper.py", line 117, in c_to_p
    var_out = super(AssemblyMapper, self).c_to_p(var_c)
  File "/Users/jfreidin/.pyenv/versions/pipecycle/lib/python2.7/site-packages/hgvs/variantmapper.py", line 347, in c_to_p
    self._validator.validate(var_c)
  File "/Users/jfreidin/.pyenv/versions/pipecycle/lib/python2.7/site-packages/hgvs/validator.py", line 33, in validate
    return self._ivr.validate(var, strict) and self._evr.validate(var, strict)
  File "/Users/jfreidin/.pyenv/versions/pipecycle/lib/python2.7/site-packages/hgvs/validator.py", line 72, in validate
    (res, msg) = self._ref_is_valid(var)
  File "/Users/jfreidin/.pyenv/versions/pipecycle/lib/python2.7/site-packages/hgvs/validator.py", line 96, in _ref_is_valid
    var_x = self.vm.c_to_n(var) if var.type == "c" else var
  File "/Users/jfreidin/.pyenv/versions/pipecycle/lib/python2.7/site-packages/hgvs/variantmapper.py", line 300, in c_to_n
    self._replace_reference(var_n)
  File "/Users/jfreidin/.pyenv/versions/pipecycle/lib/python2.7/site-packages/hgvs/variantmapper.py", line 394, in _replace_reference
    seq = self.hdp.get_seq(var.ac, pos.start.base - 1, pos.end.base)
  File "/Users/jfreidin/.pyenv/versions/pipecycle/lib/python2.7/site-packages/hgvs/decorators/lru_cache.py", line 186, in wrapper
    result = user_function(*args, **kwds)
  File "/Users/jfreidin/.pyenv/versions/pipecycle/lib/python2.7/site-packages/hgvs/dataproviders/uta.py", line 230, in get_seq
    return self.seqfetcher.fetch_seq(ac, start_i, end_i)
  File "/Users/jfreidin/.pyenv/versions/pipecycle/lib/python2.7/site-packages/hgvs/dataproviders/seqfetcher.py", line 55, in fetch_seq
    raise HGVSDataNotAvailableError("No sequence available for {ac} ({ex})".format(ac=ac, ex=ex))
HGVSDataNotAvailableError: No sequence available for NM_001429.3 (No sequence available for NM_001429.3)

**Comment by reece** [comment]
I've never heard of sporadic errors like this. A few questions:

#. You said that you're using the docker images. UTA, SeqRepo, or both?
#. How are you configuring UTA and SeqRepo?
#. Do you have logging enabled at INFO level? If not, please do so and send a line that matches (regexp) "Using.+sequence fetching". 

If you don't already have logging enabled, it's easy. Add this to your scripts:
```python
import logging
logging.basicConfig(level="INFO")
```

You should be able to run multiple processes that use hgvs at the same time (or threaded).

With the information given, my best guess is that hgvs is using a remote sequence fetching and that NCBI is rejecting rapid queries. Just a hunch. More information will help.

**Comment by jfreidin** [comment]
Hi Reece, thanks so much for your reply.

  1. I'm running a local UTA in a local Docker container.
    hgvs is talking to that. I'm a new user of this package and don't know what SeqRepo is.
    I presume I'm using a remote for that. I'd like to make that local as well.
  2. Only setting: `UTA_DB_URL=postgresql://anonymous@localhost:15032/uta/uta_20171026`
  3. I don't have logging at INFO enabled.
      I'll give that a try, but I suspect that you've already identified that SeqRepo is remote, and could be the root cause.

I was getting messages `OperationalError: FATAL:  remaining connection slots are reserved for non-replication superuser connections.`
I solved that by closing hdp after each use.
This emits an unwanted message every time:
`WARNING:hgvs.dataproviders.uta:Closing connection; future mapping and validation will fail.`

I suspect there's a better way to pool these connections but haven't had time to investigate that.

**Comment by reece** [comment]
Hi @jfreidin-

hgvs needs two kinds of data: transcript structures and sequences. UTA provides the former, whether remote or local. For sequences, your options are to use NCBI via their eutils web service (and incur the network penalty) or to use SeqRepo, a local sequence repository. Installing SeqRepo locally can be worth ~40x speed increase depending on activities. See https://hgvs.readthedocs.io/en/stable/installation.html#installing-seqrepo-optional. (Please let me know how to make this option clearer to users. Or, better, submit a documentation PR.)

You have two errors with distinct causes: 
* `OperationalError: FATAL: remaining connection slots are reserved for non-replication superuser connections.` This error results from having too many postgresql connections open. How many concurrent hgvs processes are you running?

* `HGVSDataNotAvailableError: No sequence available...` Given what you've told me, my money is on NCBI throttling your requests. Installing SeqRepo should eliminate that problem.

Finally: I removed the warning about closing the data connection in a recent release. Upgrading will make that go away.

**Comment by jfreidin** [comment]
Hi @reece,
I've installed the local SeqRepo and that's working really well.
The problem I observed was likely either at NCBI or in communication.
Sorry I saw that it was an optional step so I skipped it without recognizing the fragility and performance implications.
I'm creating a lot of objects that had their own connections because when the connections were shared across processes (via fork) they no longer worked. Rather than figure out how to arrange for one connection or connection pool/process I decided to allocate a connection for each object.
Is there a good abstraction for using hdp from multiple objects within multiprocessing.Pool subprocesses?
I'm happy to close this issue as the original problem has been addressed.
Many thanks!
